### DISCUSSION TITLE
WASM application under CCF

### DISCUSSION THREAD

**Discussion by TheBankster** [microsoft/CCF#4097] [Q&A]
I have an application I'd like to host inside CCF but it is written in Golang and compiles to WASM.
Is there any prior art I should be aware of? What is the best way to proceed?

**Comment by eddyashton** [comment]
We currently have no builtin support for executing WASM in CCF. I think there are 3 approaches to support this:

- Build a new application like `jsgeneric`, which interprets WASM rather than JavaScript. The ground truth is that CCF supports applications written in C++, but that application could be an interpreter for many other languages, particularly scripting languages (JavaScript, Lua), or bytecode (WASM). So this would be an application written in C++, ideally using some existing embeddable WASM interpreter library, and somehow exposing dynamically loaded WASM code as callable HTTP endpoints, storing state in the CCF KV.
- Use a WASM interpreter library in JS. Similar to the first suggestion, but trading off lower performance (doubly-interpreted code) for an easier implementation. This may fit well where the WASM code is purely functional, and then exposing REST endpoints and storing results in the KV can be done with the existing JS API.
- Wait for broader language support from #4062. We have plans to support application code in any languages, running in a separate trusted container. That's in-progress work, and we don't have an ETA for it being production-ready.
